aggregator/src/core.rs

```diff
@@ -62,3 +90,711 @@ pub(crate) fn extract_accumulators_and_proof(
+                    &mut offset,
+                )?;
+
+                for i in 0..4 {
+                    for j in 0..8 {
+                        // sanity check
+                        assert_exist(
+                            &batch_pi_hash_preimage[i * 8 + j + CHUNK_DATA_HASH_INDEX],
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j],
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j + 32],
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j + 64],
+                        );
+                        // assert
+                        // batch_pi_hash_preimage[i * 8 + j + CHUNK_DATA_HASH_INDEX]
+                        // = flag1 * potential_batch_data_hash_digest[(3 - i) * 8 + j]
+                        // + flag2 * potential_batch_data_hash_digest[(3 - i) * 8 + j + 32]
+                        // + flag3 * potential_batch_data_hash_digest[(3 - i) * 8 + j + 64]
+
+                        let rhs = rlc_config.mul(
+                            &mut region,
+                            &flag1,
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j],
+                            &mut offset,
+                        )?;
+                        let rhs = rlc_config.mul_add(
+                            &mut region,
+                            &flag2,
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j + 32],
+                            &rhs,
+                            &mut offset,
+                        )?;
+                        let rhs = rlc_config.mul_add(
+                            &mut region,
+                            &flag3,
+                            &potential_batch_data_hash_digest[(3 - i) * 8 + j + 64],
+                            &rhs,
+                            &mut offset,
+                        )?;
+
+                        region.constrain_equal(
+                            batch_pi_hash_preimage[i * 8 + j + CHUNK_DATA_HASH_INDEX].cell(),
+                            rhs.cell(),
+                        )?;
+                    }
+                }
+
+                // 3 batch_data_hash and chunk[i].pi_hash use a same chunk[i].data_hash when
+                // chunk[i] is not padded
+                //
+                // batchDataHash = keccak(chunk[0].dataHash || ... || chunk[k-1].dataHash)
+                //
+                // chunk[i].piHash =
+                //     keccak(
+                //        &chain id ||
+                //        chunk[i].prevStateRoot ||
+                //        chunk[i].postStateRoot ||
+                //        chunk[i].withdrawRoot  ||
+                //        chunk[i].datahash)
+                let mut randomness = Fr::default();
+                challenges.keccak_input().map(|x| randomness = x);
+                let challenge_cell =
+                    rlc_config.load_private(&mut region, &randomness, &mut offset)?;
```

Comment:
QQ: how to constrain this challenge_cell matches the actual challenge?

Reply:
75eee68463ba7891539d6dd6da6625057ef9a18f implements the solution suggested by @kunxian-xia